### ISSUE TITLE
My thoughts on git as a binary storage format

### ISSUE THREAD

**Issue by cgwalters** [docker/libpack#14]
I use git myself to store binary data such as my http://keepass.info/ database.  It works just fine for small files like that.

Beyond that though, whether this makes sense comes down to a several considerations:
 - How big is your data?
 - How often does it change?
 - When it does change, does it change in a way that a textual diff or xdelta are useful?
 - Do you need extended metadata such as uid/gid/xattrs?
 - Do you need the full history forever?
 - If serving on the network, are you ok with a "smart server"?

The big/often/change concerns all come together with x86_64 ELF, particularly things like 20MB+ statically linked golang binaries.  xdelta is just going to be a waste of CPU time on it (for OSTree, we're investigating Courgette: https://bugzilla.gnome.org/show_bug.cgi?id=721799#c27).

It looks like you're solving the metadata issue by having separate files, which I think should work just fine.  It does mean that many operations like `git checkout` and `git diff` aren't useful standalone, and will need to be wrapped.  Which isn't a big deal.

BTW, see also https://github.com/bup/bup

On the history topic, while some people may be OK keeping all of their binaries forever, conceptually they're a cache of the source code, and in a continuous integration scenario you do end up wanting garbage collection of unnecessary images.  Sometimes the simplest way to do this is to have separate dev/prod data sets.  In the dev case though, it's nice to be able to filter the data without tossing everything, something like a backup style rolling history.  Implementing this with git means you're going much lower level and against its core design.

Finally, the "smart server" consideration is a very big deal.  It locks you out of using things like CDNs/S3. And also even if you do set one up, the pack files it synthesizes are basically just going to be "tar of new objects" due to the above issue of diff/xdelta on ELF.

**Comment by cgwalters** [comment]
(Also, a notable benefit of a new format is being able to s/SHA1/SHA256 as OSTree does)

**Comment by wking** [comment]
On Fri, Jan 09, 2015 at 11:40:39AM -0800, Colin Walters wrote:
> The big/often/change concerns all come together with x86_64 ELF,
> particularly things like 20MB+ statically linked golang binaries.
> xdelta is just going to be a waste of CPU time on it (for OSTree,
> we're investigating Courgette:
> https://bugzilla.gnome.org/show_bug.cgi?id=721799#c27).

Another alternative is to use something like IPFS's ‘list’ object to
compose files from a list of sub-file blobs [1,2].  Then you can use
whatever algorithm you like to break your large binary up into
reusable chunks.  In any case, disabling compression in Git is no
worse than what we're already doing for tarballs in the registry ;).

> On the history topic, while some people may be OK keeping all of
> their binaries forever, conceptually they're a cache of the source
> code, and in a continuous integration scenario you do end up wanting
> garbage collection of unnecessary images.

I'd avoid this issue by having images be commits [3], and then Git's
usual garbage collection will remove them after they've been untagged
for a sufficient period of time.

> Finally, the "smart server" consideration is a very big deal.  It
> locks you out of using things like CDNs/S3.

You can serve Git from static HTTP, but maybe this is what you mean
here:

> And also even if you do set one up, the pack files it synthesizes
> are basically just going to be "tar of new objects" due to the above
> issue of diff/xdelta on ELF.

So don't create the packs for the layer tarballs.  Or break down the
binaries like IPFS and bup.  In any case, I'm not opposed to writing
our own transport if we can't use Git's native protocols out of the
box.

[1]: https://github.com/jbenet/ipfs/blob/f05ac1b41666b0134298fd1fe56d95c02a6ddb11/papers/ipfs-cap2pfs/ipfs-cap2pfs.tex#L748
[2]: http://static.benet.ai/t/ipfs.pdf
[3]: https://groups.google.com/a/dockerproject.org/d/msg/distribution/TP9V-oUJ_vA/PYEyDq-M0EkJ

**Comment by wking** [comment]
On Fri, Jan 09, 2015 at 11:40:39AM -0800, Colin Walters wrote:
> for OSTree, we're investigating Courgette:
> https://bugzilla.gnome.org/show_bug.cgi?id=721799#c27

Courgette looks nice, but as far as I can see the patent issues [1]
are unresolved [2,3] (as of 2013-09-24, as far as Jean-Marc could see
;).

[1]: http://lwn.net/Articles/359939/
[2]: https://bugzilla.mozilla.org/show_bug.cgi?id=504624#c9
[3]: https://bugzilla.mozilla.org/show_bug.cgi?id=504624#c34

**Comment by wking** [comment]
On Fri, Jan 09, 2015 at 11:43:59AM -0800, Colin Walters wrote:
> (Also, a notable benefit of a new format is being able to
> s/SHA1/SHA256 as OSTree does)

This would be nice, especially since the security of Git's SHA1
doesn't hold up as well once you start using it to pass around binary
data.

**Comment by shykes** [comment]
I really like IPFS, and I'm talking to its author about suitability for this. But so far it seems pretty young so I'm not sure I'm ready to adopt it as a standard.

On the security of SHA1: I am fully aware of that, regular git hashes should never be used outside the trust domain, ever. However that is not incompatible with using regular git *inside* the trust domain, and layering a more secure hash when packaging and distributing data across trust domains. That way you get the best of both worlds: within your trust domain (a single host, or perhaps a cluster of hosts) you benefit from the tooling and aggregate engineering effort behind git. For example you could drop down to the git command line to inspect a database at any time. And you still get to add things which git doesn't do (like security), cleanly layered on top.

On git's support for extended attributes: that is not a problem either. I don't plan on relying on git for that. You're correct, 'checkout' and 'commit' won't work out of the box for that, but it's an acceptable tradeoff.

On needing the full history: I plan on keeping git's history feature an implementation detail. So, whether the full history is kept forever or not is up to the implementation, not to the user. That is a marked difference with how people use git for source control. libpack uses history more as a "bin log": something which is bounded by a window at the discretion of the software.

On bup: I am very familiar with it. They take a different approach to the problem, since I believe they keep the tar stream as a single file. I prefer to split up the tar archive to take advantage of git trees.

On the performance issues of packing large files: I am not married to git's packing and index format. I think there should be a gateway to using it, for fuller compatibility with the git tooling, but it must be possible to rip it out and insert a different storage and transport backend. It's perfectly fine to do this while keeping git's data model. For example there is no performance problem whatsoever when objects are in loose storage.

On "smart server": same as above. You can use git's data model without forcing everyone to use regular "git push" (but you can still make that available which is nice).

**Comment by cgwalters** [comment]
On Fri, 2015-01-09 at 12:10 -0800, W. Trevor King wrote:
> On Fri, Jan 09, 2015 at 11:40:39AM -0800, Colin Walters wrote: 
> > The big/often/change concerns all come together with x86_64 ELF, 
> > particularly things like 20MB+ statically linked golang binaries. 
> > xdelta is just going to be a waste of CPU time on it (for OSTree, 
> > we're investigating Courgette: 
> > https://bugzilla.gnome.org/show_bug.cgi?id=721799#c27). 
> 
> Another alternative is to use something like IPFS's ‘list’ object to 
> compose files from a list of sub-file blobs [1,2]. Then you can use 
> whatever algorithm you like to break your large binary up into 
> reusable chunks.

Concatenation isn't good enough; ELF has the "cascading change"
property that's similar to anything with compression, where a change
in one portion of the file can cause rolling byte-level changes
elsewhere.

See the Courgette rationale.

Anyways, in a lot of cases, "only send/store the new files" is a
reasonable enough implementation, particularly if you have the ability
to influence the input file format.

For example, say you have a collaboratively edited Theora/MPEG video.
If it's OK for the people using it to have it broken up into 30
minute/whatever chunks, then content addressed storage gets more
palatable.

(Of course, this makes the video larger, and this is a general theme:
 compression makes deltas harder, and vice versa)

> I'd avoid this issue by having images be commits [3], and then Git's 
> usual garbage collection will remove them after they've been untagged 
> for a sufficient period of time. 

"shipped to prod and need to be reproducible images be commits" you
mean?  Sure, but probably saner to structure it as shipped images
are tags, and to GC commits. 

> You can serve Git from static HTTP, but maybe this is what you mean 
> here: 

Right.  It's what OSTree does, and it's either slow (one http request
per file as the current OSTree), or forces clients to retrieve large
amounts of history (pack files).  We're working on a custom delta
format that tries to fit in between these two extremes.

**Comment by wking** [comment]
On Fri, Jan 09, 2015 at 12:44:52PM -0800, Colin Walters wrote:
> For example, say you have a collaboratively edited Theora/MPEG
> video.  If it's OK for the people using it to have it broken up into
> 30 minute/whatever chunks, then content addressed storage gets more
> palatable.

Just to be clear, IPFS's ‘list’ is purely a storage-side construct.  A
list (or tree of nested lists) gets extracted to the filesystem as a
single file.  On the other hand, I'm not sure how static chunks of
video are in the face of editing.  I expect editors try to avoid
re-encoding uneditted chunks, but they might get re-framed, which
would dirty all the blobs in your list (similar to the reference issue
Courgette works with).

> > I'd avoid this issue by having images be commits [3], and then
> > Git's usual garbage collection will remove them after they've been
> > untagged for a sufficient period of time.
> 
> "shipped to prod and need to be reproducible images be commits" you
> mean?

I'm having trouble parsing that ;).  Can you rephrase it for me?

> Sure, but probably saner to structure it as shipped images are tags,
> and to GC commits.

I'm suggesting we use tags for signatures [1].  That means once you
untag an image commit, Git is free to garbage-collect it.

> > You can serve Git from static HTTP, but maybe this is what you
> > mean here:
> 
> Right.  It's what OSTree does, and it's either slow (one http
> request per file as the current OSTree), or forces clients to
> retrieve large amounts of history (pack files).  We're working on a
> custom delta format that tries to fit in between these two extremes.

I think it's going to be hard to do better than rolling checksums
(like bup) for this, unless you have time to plug in a fancier delta
format for each class of binary-dirting changes (ELF references, video
frame metadata, …).  But if you get something working, that would be
awesome :).

[1]: https://groups.google.com/a/dockerproject.org/forum/#!msg/distribution/TP9V-oUJ_vA/PYEyDq-M0EkJ

**Comment by shykes** [comment]
Hi all, I think we are drifting away from the scope of a github issue on the libpack repo. I suggest taking this conversation to the newly created distribution@dockerproject.org mailing list, where we will discuss all things packaging & distribution for Docker-land.

cc @icecrime who set up the mailing list.

**Comment by wking** [comment]
On Fri, Jan 09, 2015 at 12:42:56PM -0800, Solomon Hykes wrote:
> On the security of SHA1: I am fully aware of that, regular git
> hashes should never be used outside the trust domain, ever. However
> that is not incompatible with using regular git *inside* the trust
> domain, and layering a more secure hash when packaging and
> distributing data across trust domains. That way you get the best of
> both worlds: within your trust domain (a single host, or perhaps a
> cluster of hosts) you benefit from the tooling and aggregate
> engineering effort behind git. For example you could drop down to
> the git command line to inspect a database at any time. And you
> still get to add things which git doesn't do (like security),
> cleanly layered on top.

The hash used (SHA1 for Git) is fairly central to Git's object
linkage.  I understand how you could repace it with a different hash,
but not how you could layer a different hash on top.  Can you explain
how that would work?  What does a tree look like in your layered
system?  Do you have one tree object with entries like:

  100644 blob <SHA1> <OTHER-HASH> <filename>

?  Do you have two tree objects for each node, one in the SHA1 store
with SHA1 hashes and the other in the alternate store with alternate
hashes?

> On bup: I am very familiar with it. They take a different approach
> to the problem, since I believe they keep the tar stream as a single
> file. I prefer to split up the tar archive to take advantage of git
> trees.

I think a Docker-distributing tool could go either way here.  Using
trees instead of tarballs would be nice, but either way I don't think
it would be a problem for adopting a Git-based tool.

**Comment by cgwalters** [comment]
I'm not finding the distribution list; it's not linked from https://dockerproject.com/ and my initial guess of https://groups.google.com/forum/#!forum/docker-distribution appears to ENOENT.

**Comment by icecrime** [comment]
@cgwalters Sorry, [this is the correct link](https://groups.google.com/a/dockerproject.org/forum/#!forum/distribution)

**Comment by shykes** [comment]
Just to re-iterate my previous comment: I think this is a very valuable discussion but we should have it [here](https://groups.google.com/a/dockerproject.org/forum/#!forum/distribution) so that it is not tied to a particular implementation.

Thanks everyone for moving this topic forward!
